### PR TITLE
Prevent JSOC tests from being run in parallel

### PR DESCRIPTION
**Pull request by dstansby**
Today's CI failure was
```python
E       drms.exceptions.DrmsExportError: User jsoc@cadair.com has 1 pending export requests (JSOC_20220121_387); please wait until at least one request has completed before submitting a new one. [status=7]

../../.tox/py38-online/lib/python3.8/site-packages/drms/client.py:185: DrmsExportError
```

This PR:
- Uses `pytest-xdist` by default, with the auto option (this was already being used on CI)
- Marks tests that download JSOC data as being in the same group. This ensures they are allocated the same worker and are run in serial, hopefully avoiding pending export errors.

I don't think we should backport a new test dependency, but am open to other opinions on that.

### CODE REVIEW

**Review by ConorMacBride — COMMENTED** [review]
Makes sense to ensure that two requests are not made before another finishes. Should we also stagger the daily scheduled runs to reduce the chance of this happening across runs?

I'm not sure about making parallel runs the default. Output from a parallel run is (slightly) more difficult to follow and, I haven't tested it, but maybe `breakpoint()` won't work in that case. It should be possible to [dynamically include](https://docs.pytest.org/en/6.2.x/example/simple.html#dynamically-adding-command-line-options) `--dist loadgroup` in `conftest.py` if `pytest-xdist` is installed and `-n` is set.

We can add a helper function that only applies the `xdist_mark` group if xdist is actually installed. That way we can backport the groups without adding xdist as a dependency. Although, I think it makes sense to add `pytest-xdist` as a dependency, and for the same reasons, maybe we should make `tox` a dependency also?

**Comment by nabobalis** [comment]
We can add pytest-xdist as a test dep going backwards.

**Comment by dstansby** [comment]
If someone else wants to take on backporting this feel free. Note that this will involve checking if dependencies need to be updated for wheel building/testing and conda building. If something needs updating with conda please open an issue on the conda-forge repo so I can update it at the next release.

> Makes sense to ensure that two requests are not made before another finishes. Should we also stagger the daily scheduled runs to reduce the chance of this happening across runs?

👍 

> I'm not sure about making parallel runs the default. Output from a parallel run is (slightly) more difficult to follow and, I haven't tested it, but maybe `breakpoint()` won't work in that case. It should be possible to [dynamically include](https://docs.pytest.org/en/6.2.x/example/simple.html#dynamically-adding-command-line-options) `--dist loadgroup` in `conftest.py` if `pytest-xdist` is installed and `-n` is set.

👍 - parallel is now disabled by default. I took the easy route of manually putting --dist loadgroup in the pipelines configuration.

> We can add a helper function that only applies the `xdist_mark` group if xdist is actually installed. That way we can backport the groups without adding xdist as a dependency. Although, I think it makes sense to add `pytest-xdist` as a dependency, and for the same reasons, maybe we should make `tox` a dependency also?

I think this is too much of a hassle.

File: sunpy/net/jsoc/tests/test_jsoc.py [hunk 1]
```diff
@@ -15,6 +15,9 @@
 from sunpy.net.jsoc import JSOCClient, JSOCResponse
 from sunpy.util.exceptions import SunpyUserWarning
 
+# Ensure all JSOC tests are run on the same parallel worker
+pytest.mark.xdist_group(name="jsoc")
```
Comment (ConorMacBride) [inline sunpy/net/jsoc/tests/test_jsoc.py:19]: See [pytest docs](https://docs.pytest.org/en/6.2.x/example/markers.html#marking-whole-classes-or-modules). Running `pytest --co -m "xdist_group"` shows this marks the whole module.
```suggestion
pytestmark = pytest.mark.xdist_group(name="jsoc")
```

**Review by ConorMacBride — CHANGES_REQUESTED** [review]
Just one update for the module level mark.

When the figure tests were using `-n auto` they used 18 threads and a thread crashed, failing the tests. So we should probably wait for #5830 to be merged so we can test Azure with `-n auto` (or just change `auto` to `4`?). The Core Tests stage is still running with `-n=4` which is probably fine.

The conda recipe will need `pytest-xdist` due to the `--dist` option and the `xdist_group` mark. I'll open an issue there.

**Review by ConorMacBride — APPROVED** [review]

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
> Makes sense to ensure that two requests are not made before another finishes. Should we also stagger the daily scheduled runs to reduce the chance of this happening across runs?

You want to open a PR changing this?

**Comment by nabobalis** [comment]
Do we also want to change the core tests to use -n auto too?

**Comment by ConorMacBride** [comment]
Yeah, I can open a PR for the scheduled runs. I can't see any difference in the times with 2 vs 4. Maybe just change all to auto for consistency?

**Comment by nabobalis** [comment]
Works for me.

File: setup.cfg [hunk 2]
```diff
@@ -109,7 +110,7 @@ testpaths = "sunpy" "docs"
 norecursedirs = ".tox" "build" "docs[\/]_build" "docs[\/]generated" "*.egg-info" "examples" "sunpy[/\]_dev" ".jupyter" ".history" "tools" "sunpy[\/]extern" ".pyinstaller"
 doctest_plus = enabled
 doctest_optionflags = NORMALIZE_WHITESPACE FLOAT_CMP ELLIPSIS
-addopts = --doctest-rst --doctest-ignore-import-errors -p no:unraisableexception -p no:threadexception
+addopts = --dist no --doctest-rst --doctest-ignore-import-errors -p no:unraisableexception -p no:threadexception
```
Comment (Cadair) [inline setup.cfg:113]: Why did you add `--dist no` here? What's it achieving that isn't by default? Don't you have to specify `-n` to enable pytest-xdist?

Comment (Cadair) [reply setup.cfg:113]: Why do we want to ensure pytest-xdist is installed? To be clear I don't care that it's here, I just want to understand why.

Comment (nabobalis) [reply setup.cfg:113]: Not a clue. At least this way, it ensures its installed and listed as a test deps instead of only working via tox/CI.
